src/main/java/rx/internal/operators/OperatorConcat.java

```diff
@@ -172,7 +172,8 @@ void subscribeNext() {
                     child.onCompleted();
                 } else if (o != null) {
                     Observable<? extends T> obs = nl.getValue(o);
-                    currentSubscriber = new ConcatInnerSubscriber<T>(this, child, requested);
+                    currentSubscriber = new ConcatInnerSubscriber<T>(this, child);
+                    currentSubscriber.requestMore(requested);
```

Comment:
This doesn't really help much (the constructor version is behaviorally equivalent) because there is still a small window between requesting the ```requested``` and setting the ```current```. Swapping the ```requestMore``` and ```set``` doesn't help either because then a concurrent get will request r0 + 2n: r0 + n because requested is updated and n again because the reference is non-null. You need my ```ProducerArbiter```, however, If you can wait a few hours, I'll post a PR with all my useful producers put into ```rx.internal.producers``` since making them 'official' takes several days of back-and-forth.